backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,326 @@
+                                        };
+
+                                        var i = tmpCIAttributes.FindIndex(e => e.Name == attribute.Name);
+
+                                        if (i != -1)
+                                        {
+                                            // check if current attribute has higher priority
+                                            if (tmpCIAttributes[i].Priority < a.Priority)
+                                            {
+                                                tmpCIAttributes.RemoveAt(i);
+                                                tmpCIAttributes.Add(a);
+                                            }
+                                        } else
+                                        {
+                                            tmpCIAttributes.Add(a);
+                                        }
+
+                                        attribute.Name = a.Name;
+                                        attribute.Value = a.Value;
+                                    }
+                                }
+                            }
+
+                            gatheredAttributes.AddRange(tmpCIAttributes);
+                        }
+                    }
+
+                    // for all gathered attributes insert them to the main ci
+
+                    foreach (var mapping in cfg.BaseCI.AttributeMapping)
+                    {
+
+                        foreach (var attribute in gatheredAttributes)
+                        {
+                            // first we need to check input whitelist and blacklist
+
+                            if (!IsAttributeAllowed(attribute.Name, cfg.BaseCI.InputWhitelist, cfg.BaseCI.InputBlacklist))
+                            {
+                                continue;
+                            }
+
+                            // check base ci attribute mapping
+
+                            if (!IsAttributeIncludedInSource(attribute.Name, mapping.Source))
+                            {
+                                continue;
+                            }
+
+                            // check if attribute exists in base ci
+
+                            if (mainCI.MergedAttributes.ContainsKey(attribute.Name) && ciPrio < attribute.Priority)
+                            {
+                                // don't change this attribute since 
+                                continue;
+                            }
+
+                            var (_, changed) = await attributeModel.InsertAttribute(
+                                GetTargetName(attribute.Name, mapping.Target),
+                                new AttributeScalarValueText(attribute.Value),
+                                mainCI.ID,
+                                layer,
+                                changesetProxy,
+                                new DataOriginV1(DataOriginType.Manual),
+                                trans);
```

Comment:
Inserting the generated attributes like this has problems:

- it's slow, because you are writing one attribute at a time
- you are never deleting any old attributes, which is definitely not correct
- you are writing the attributes into the same layer as you are reading them from, which is not good and not the intention. Instead, you should write all generated attributes into `targetLayer` (the layer ID you get as a parameter, which is the layer you bind the CLB to)
- DataOrigin should be `ComputeLayer`

It'd be much better if you used `attributeModel.BulkReplaceAttributes()`, kinda like this:
```
await attributeModel.BulkReplaceAttributes(new BulkCIAttributeDataLayerScope("", targetLayer, attributeFragments), changesetProxy, new DataOriginV1(DataOriginType.ComputeLayer), trans);
```